### PR TITLE
DASD-4403 Azure Search, Cosmos DB and Keyvault secrets ARM templates

### PR DESCRIPTION
**Pull request by rivoric**
* Added Azure Search and Cosmos DB templates with full test.
* Added a companion template for keyvaults. Not easy to automate the test but has been manually tested.

### CODE REVIEW

File: ArmTemplates/azure-search.json [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            "type": "string",
+            "minLength": 2,
+            "maxLength": 60,
+            "metadata": {
+                "description": "Service name must only contain lowercase letters, digits or dashes, cannot use dash as the first two or last one characters, cannot contain consecutive dashes, and is limited between 2 and 60 characters in length."
+            }
+        },
+        "azureSearchSku": {
+            "type": "string",
+            "defaultValue": "standard",
```
Comment (NickGraham101) [inline ArmTemplates/azure-search.json]: I think default should be basic.  We've load tested at 20 searches per second on basic.  That said you can't scale from basic to standard which is an argument to default to standard.  If we're creating lots of small search services basic will be a lot more cost effective ~£50 against ~£180 per month

File: ArmTemplates/azure-search.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Azure Search
+
+Creates an Azure Search
+
+## Paramaters
+
+azureSearchName: (required) string
+
+Name of key vault. Will be created in the same resource group as the script is run and in the default location for resource group.
```
Comment (NickGraham101) [inline ArmTemplates/azure-search.md]: Name of search instance

File: ArmTemplates/azure-search.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+## Paramaters
+
+azureSearchName: (required) string
+
+Name of key vault. Will be created in the same resource group as the script is run and in the default location for resource group.
+
+azureSearchSku: (optional) string
+
+Can be either free, basic, standard, standard2 or standard3.
+Will default to standard if not supplied.
```
Comment (NickGraham101) [inline ArmTemplates/azure-search.md]: Note:  you cannot scale from basic to standard.

File: ArmTemplates/keyvault-secrets.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
+# KeyVault secrets
+
+Creates secrets within the key vault
+
+## Paramaters
+
+keyVaultName: (required) string
+
+Name of key vault. Will be created in the same resource group as the script is run and in the default location for resource group.
```
Comment (NickGraham101) [inline ArmTemplates/keyvault-secrets.md]: KeyVault must already exist or be defined separately in a template linked to or nested within the same parent.  Secrets will be created in this KeyVault.

File: Tests/A005.ArmTemplates.azure-search.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+      TemplateParameterObject = $TemplateParameters
+    }
+
+    $output = Test-AzureRmResourceGroupDeployment @TestTemplateParams
+  
+    It "Should be deployed successfully" {
+      $output | Should -Be $null
+    }
+
+  }
```
Comment (NickGraham101) [inline Tests/A005.ArmTemplates.azure-search.Tests.ps1]: Should test with non-default values as well.  (I've not ran this)

  Context "When Azure Search is deployed with just search name" { 
    $TemplateParameters = @{ 
      azureSearchName = "dfc-foo-bar-sch" 
      azureSearchReplicaCount = 2
      azureSearchPartitionCount = 2
    } 
    $TestTemplateParams = @{ 
      ResourceGroupName       = $ResourceGroupName 
      TemplateFile            = $TemplateFile 
      TemplateParameterObject = $TemplateParameters 
    } 
 
    $output = Test-AzureRmResourceGroupDeployment @TestTemplateParams 
 
    It "Should be deployed successfully" { 
      $output | Should -Be $null 
    } 
 
  }

Comment (rivoric) [reply Tests/A005.ArmTemplates.azure-search.Tests.ps1]: Added a couple more tests. I've not tested every possible value of every parameter, just enough to be confident it works.

File: ArmTemplates/keyvault-secrets.md (hunk 4 above)
Comment (rivoric) [reply ArmTemplates/keyvault-secrets.md]: Done

File: ArmTemplates/azure-search.md (hunk 3 above)
Comment (rivoric) [reply ArmTemplates/azure-search.md]: Added additional comments

File: ArmTemplates/azure-search.md (hunk 2 above)
Comment (rivoric) [reply ArmTemplates/azure-search.md]: Copy and paste error :)

File: ArmTemplates/azure-search.json (hunk 1 above)
Comment (rivoric) [reply ArmTemplates/azure-search.json]: Agreed and changed

**Pull request merged**